texlive/texmf-dist/doc/man/man1/pdftex.1
.TH PDFTEX 1 "1 March 2011" "pdftex 1.40" "Web2C 2013"
.\"=====================================================================
.if n .ds MF Metafont
.if t .ds MF M\s-2ETAFONT\s0
.if t .ds TX \fRT\\h'-0.1667m'\\v'0.20v'E\\v'-0.20v'\\h'-0.125m'X\fP
.if n .ds TX TeX
.if n .ds MF Metafont
.if t .ds MF M\s-2ETAFONT\s0
.ie t .ds OX \fIT\v'+0.25m'E\v'-0.25m'X\fP for troff
.el .ds OX TeX for nroff
.\" the same but obliqued
.\" BX definition must follow TX so BX can use TX
.if t .ds BX \fRB\s-2IB\s0\fP\*(TX
.if n .ds BX BibTeX
.\" LX definition must follow TX so LX can use TX
.if t .ds LX \fRL\\h'-0.36m'\\v'-0.15v'\s-2A\s0\\h'-0.15m'\\v'0.15v'\fP\*(TX
.if n .ds LX LaTeX
.if t .ds AX \fRA\\h'-0.1667m'\\v'0.20v'M\\v'-0.20v'\\h'-0.125m'S\fP\*(TX
.if n .ds AX AmSTeX
.if t .ds AY \fRA\\h'-0.1667m'\\v'0.20v'M\\v'-0.20v'\\h'-0.125m'S\fP\*(LX
.if n .ds AY AmSLaTeX
.if n .ds WB Web
.if t .ds WB W\s-2EB\s0
.\"=====================================================================
.SH NAME
pdftex, pdfinitex, pdfvirtex \- PDF output from TeX
.SH SYNOPSIS
.B pdftex
.RI [ options ]
.RI [ \fB&\fPformat ]
.RI [ file | \fB\e\fPcommands ]
.\"=====================================================================
.SH DESCRIPTION
Run the pdf\*(TX typesetter on
.IR file ,
usually creating
.IR file.pdf .
If the file argument has no extension, ".tex" will be appended to it. 
Instead of a filename, a set of pdf\*(TX commands can be given, the first
of which must start with a backslash.
With a 
.BI & format
argument pdf\*(TX uses a different set of precompiled commands,
contained in
.IR format\fB.fmt\fP ;
it is usually better to use the
.B -fmt
.I format
option instead.
.PP
pdf\*(TX is a version of \*(TX, with the e-\*(TX extensions, that can create
.I PDF
files as well as
.I DVI
files.
.PP
In
.I DVI
mode, pdf\*(TX can be used as a complete replacement for the \*(TX
engine.
.PP
The typical use of pdf\*(TX is with a pregenerated formats for which
PDF output has been enabled.  The
.B pdftex
command uses the equivalent of the plain \*(TX format, and the
.B pdflatex
command uses the equivalent of the \*(LX format.
To generate formats, use the
.B -ini
switch.
.PP
The
.B pdfinitex
and
.B pdfvirtex
commands are pdf\*(TX's analogues to the
.B initex
and
.B virtex
commands.  In this installation, if the links exist, they are symbolic
links to the
.B pdftex
executable.
.PP
In
.I PDF
mode, pdf\*(TX can natively handle the
.IR PDF ,
.IR JPG ,
.IR JBIG2 ,
and
.I PNG
graphics formats.  pdf\*(TX cannot include PostScript or Encapsulated
PostScript (EPS) graphics files; first convert them to PDF using 
.BR epstopdf (1).
pdf\*(TX's handling of its command-line arguments is similar to that of
of the other \*(TX programs in the
.I web2c
implementation.
.\"=====================================================================
.SH OPTIONS
This version of pdf\*(TX understands the following command line options.
.TP
.B -draftmode
Sets \epdfdraftmode so pdf\*(TX doesn't write a PDF and doesn't read any
included images, thus speeding up execution.
.TP
.B -enc
Enable the enc\*(TX extensions.  This option is only effective in
combination with
.BR -ini .
For documentation of the enc\*(TX extensions see
.BR http://www.olsak.net/enctex.html .
.TP
.B -etex
Enable the e-\*(TX extensions.  This option is only effective in
combination with
.BR -ini .
See
.BR etex (1).
.TP
.B -file-line-error
Print error messages in the form
.I file:line:error
which is similar to the way many compilers format them.
.TP
.B -no-file-line-error
Disable printing error messages in the
.I file:line:error
style.
.TP
.B -file-line-error-style
This is the old name of the
.B -file-line-error
option.
.TP
.BI -fmt \ format
Use
.I format
as the name of the format to be used, instead of the name by which
pdf\*(TX was called or a
.I %&
line.
.TP
.B -halt-on-error
Exit with an error code when an error is encountered during processing.
.TP
.B -help
Print help message and exit.
.TP
.B -ini
Start in
.I INI
mode, which is used to dump formats.  The
.I INI
mode can be used for typesetting, but no format is preloaded, and
basic initializations like setting catcodes may be required.
.TP
.BI -interaction \ mode
Sets the interaction mode.  The mode can be either
.IR batchmode ,
.IR nonstopmode ,
.IR scrollmode ,
and
.IR errorstopmode .
The meaning of these modes is the same as that of the corresponding
\ecommands.
.TP
.B -ipc
Send DVI or PDF output to a socket as well as the usual output file.
Whether this option is available is the choice of the installer.
.TP
.B -ipc-start
As
.BR -ipc ,
and starts the server at the other end as well.  Whether this option
is available is the choice of the installer.
.TP
.BI -jobname \ name
Use
.I name
for the job name, instead of deriving it from the name of the input file.
.TP
.BI -kpathsea-debug \ bitmask
Sets path searching debugging flags according to the bitmask.  See the
.I Kpathsea
manual for details.
.TP
.BI -mktex \ fmt
Enable
.RI mktex fmt ,
where
.I fmt
must be either
.I tex
or
.IR tfm .
.TP
.B -mltex
Enable ML\*(TX extensions.  Only effective in combination with
.BR -ini .
.TP
.BI -no-mktex \ fmt
Disable
.RI mktex fmt ,
where
.I fmt
must be either
.I tex
or
.IR tfm .
.TP
.BI -output-comment \ string
In
.I DVI
mode, use
.I string
for the
.I DVI
file comment instead of the date.  This option is ignored
in
.I PDF
mode.
.TP
.BI -output-directory \ directory
Write output files in
.I directory
instead of the current directory.  Look up input files in
.I directory
first, the along the normal search path.
.TP
.BI -output-format \ format
Set the output format mode, where
.I format
must be either
.I pdf
or
.IR dvi .
This also influences the set of graphics formats understood by pdf\*(TX.
.TP
.B -parse-first-line
If the first line of the main input file begins with
.I %&
parse it to look for a dump name or a
.B -translate-file
option.
.TP
.B -no-parse-first-line
Disable parsing of the first line of the main input file.
.TP
.BI -progname \ name
Pretend to be program
.IR name .
This affects both the format used and the search paths.
.TP
.B -recorder
Enable the filename recorder.
This leaves a trace of the files opened for input and output
in a file with extension
.IR .fls .
.TP
.B -shell-escape
Enable the
.BI \ewrite18{ command }
construct.  The
.I command
can be any shell command.  This construct is normally
disallowed for security reasons.
.TP
.B -no-shell-escape
Disable the
.BI \ewrite18{ command }
construct, even if it is enabled in the
.I texmf.cnf
file.
.TP
.B -src-specials
In
.I DVI
mode, insert source specials into the
.I DVI
file.
This option is ignored in
.I PDF
mode.
.TP
.BI -src-specials \ where
In
.I DVI
mode, insert source specials in certain placed of the
.I DVI
file.
.I where
is a comma-separated value list:
.IR cr ,
.IR display ,
.IR hbox ,
.IR math ,
.IR par ,
.IR parent ,
or
.IR vbox .
This option is ignored in
.I PDF
mode.
.TP
.BI -translate-file \ tcxname
Use the
.I tcxname
translation table to set the mapping of input characters and
re-mapping of output characters.
.TP
.BI -default-translate-file \ tcxname
Like
.B -translate-file
except that a
.I %&
line can overrule this setting.
.TP
.B -version
Print version information and exit.
.\"=====================================================================
.SH ENVIRONMENT
See the Kpathsearch library documentation (the `Path specifications'
node) for precise details of how the environment variables are used.
The
.B kpsewhich
utility can be used to query the values of the variables.
.PP
One caveat: In most pdf\*(TX formats, you cannot use ~ in a filename you
give directly to pdf\*(TX, because ~ is an active character, and hence is
expanded, not taken as part of the filename.  Other programs, such as
\*(MF, do not have this problem.
.PP
.TP
.B TEXMFOUTPUT
Normally, pdf\*(TX puts its output files in the current directory.  If
any output file cannot be opened there, it tries to open it in the
directory specified in the environment variable TEXMFOUTPUT.
There is no default value for that variable.  For example, if you say
.I pdftex paper
and the current directory is not writable and TEXMFOUTPUT has
the value
.IR /tmp ,
pdf\*(TX attempts to create
.I /tmp/paper.log
(and
.IR /tmp/paper.pdf ,
if any output is produced.)  TEXMFOUTPUT is also checked for input
files, as \*(TX often generates files that need to be subsequently
read; for input, no suffixes (such as ``.tex'') are added by default,
the input name is simply checked as given.
.TP
.B TEXINPUTS
Search path for
.I \einput
and
.I \eopenin
files.
This should start with ``.'', so
that user files are found before system files.  An empty path
component will be replaced with the paths defined in the
.I texmf.cnf
file.  For example, set TEXINPUTS to ".:/home/user/tex:" to prepend the
current directory and ``/home/user/tex'' to the standard search path.
.TP
.B TEXFORMATS
Search path for format files.
.TP
.B TEXPOOL
search path for
.B pdftex
internal strings.
.TP
.B TEXEDIT
Command template for switching to editor.  The default, usually
.BR vi ,
is set when pdf\*(TX is compiled.
.TP
.B TFMFONTS
Search path for font metric
.RI ( .tfm )
files.
.\"=====================================================================
.SH FILES
The location of the files mentioned below varies from system to
system.  Use the
.B kpsewhich
utility to find their locations.
.TP
.I pdftex.pool
Text file containing pdf\*(TX's internal strings.
.TP
.I pdftex.map
Filename mapping definitions.
.TP
.I *.tfm
Metric files for pdf\*(TX's fonts.
.TP
.I *.fmt
Predigested pdf\*(TX format (.\|fmt) files.
.br
.\"=====================================================================
.SH NOTES
Starting with version 1.40, pdf\*(TX incorporates the e-\*(TX
extensions, and pdfe\*(TX is just a copy of pdf\*(TX.  See 
.BR etex (1).
This manual page is not meant to be exhaustive.  The complete
documentation for this version of pdf\*(TX can be found in the 
.IR "pdf\*(TX manual"
and the info manual
.IR "Web2C: A TeX implementation" .
.\"=====================================================================
.SH BUGS
This version of pdf\*(TX implements a number of optional extensions.
In fact, many of these extensions conflict to a greater or lesser
extent with the definition of pdf\*(TX.  When such extensions are
enabled, the banner printed when pdf\*(TX starts is changed to print
.B pdfTeXk
instead of
.BR pdfTeX .
.PP
This version of pdf\*(TX fails to trap arithmetic overflow when
dimensions are added or subtracted.  Cases where this occurs are rare,
but when it does the generated
.I DVI
file will be invalid.  Whether a generated
.I PDF
file would be usable is unknown.
.\"=====================================================================
.SH AVAILABILITY
pdf\*(TX is available for a large variety of machine architectures
and operation systems.
pdf\*(TX is part of all major \*(TX distributions.
.P
Information on how to get pdf\*(TX and related information
is available at the
.B http://www.pdftex.org
.IR "pdf\*(TX"
web site.
.P
The following pdfe\*(TX related mailing list is available:
.BR pdftex@tug.org .
This is a mailman list;
to subscribe send a message containing
.I subscribe
to
.BR pdftex-request@tug.org .
A web interface and list archives can be found at the
.B "http://lists.tug.org/pdftex"
mailing list web site.
.\"=====================================================================
.SH "SEE ALSO"
.BR epstopdf (1),
.BR etex (1),
.BR latex (1),
.BR mptopdf (1),
.BR tex (1),
.BR mf (1).
http://tug.org/applications/pdftex, http://tug.org/web2c.
.\"=====================================================================
.SH AUTHORS
The primary authors of pdf\*(TX are Han The Thanh, Petr Sojka,
Jiri Zlatuska, and Peter Breitenlohner (e\*(TX).
.PP
\*(TX was designed by Donald E. Knuth,
who implemented it using his \*(WB system for Pascal programs.
It was ported to Unix at Stanford by Howard Trickey, and
at Cornell by Pavel Curtis.
The version now offered with the Unix \*(TX distribution is that
generated by the \*(WB to C system
.RB ( web2c ),
originally written by Tomas Rokicki and Tim Morgan.
.PP
The enc\*(TX extensions were written by Petr Olsak.
.\" vim: syntax=nroff


pdftex.js
var PDFTeX = function() {
  var worker = new Worker("pdftex-worker.js");
  var self = this;
  var initialized = false;

  self.on_stdout = function(msg) {
    console.log(msg);
  }

  self.on_stderr = function(msg) {
    console.log(msg);
  }


  worker.onmessage = function(ev) {
    var data = JSON.parse(ev.data);
    var msg_id;

    if(!('command' in data))
      console.log("missing command!", data);
    switch(data['command']) {
      case 'ready':
        onready.done(true);
        break;
      case 'stdout':
      case 'stderr':
        self['on_'+data['command']](data['contents']);
        break;
      default:
        //console.debug('< received', data);
        msg_id = data['msg_id'];
        if(('msg_id' in data) && (msg_id in promises)) {
          promises[msg_id].done(data['result']);
        }
        else
          console.warn('Unknown worker message '+msg_id+'!');
    }
  }

  var onready = new promise.Promise();
  var promises = [];
  var chunkSize = undefined;

  var sendCommand = function(cmd) {
    var p = new promise.Promise();
    var msg_id = promises.push(p)-1;

    onready.then(function() {
      cmd['msg_id'] = msg_id;
      //console.debug('> sending', cmd);
      worker.postMessage(JSON.stringify(cmd));
    });

    return p;
  };

  var determineChunkSize = function() {
    var size = 1024;
    var max = undefined; 
    var min = undefined;
    var delta = size;
    var success = true;
    var buf;

    while(Math.abs(delta) > 100) {
      if(success) {
        min = size;
        if(typeof(max) === 'undefined')
          delta = size;
        else
          delta = (max-size)/2;
      }
      else {
        max = size;
        if(typeof(min) === 'undefined')
          delta = -1*size/2;
        else
          delta = -1*(size-min)/2;
      }
      size += delta;

      success = true;
      try {
        buf = String.fromCharCode.apply(null, new Uint8Array(size));
        sendCommand({
          command: 'test',
          data: buf,
        });
      }
      catch(e) {
        success = false;
      }
    }

    return size;
  };


  var createCommand = function(command) {
    self[command] = function() {
      var args = [].concat.apply([], arguments);

      return sendCommand({
        'command':  command,
        'arguments': args,
      });
    }
  }
  createCommand('FS_createDataFile'); // parentPath, filename, data, canRead, canWrite
  createCommand('FS_readFile'); // filename
  createCommand('FS_unlink'); // filename
  createCommand('FS_createFolder'); // parent, name, canRead, canWrite
  createCommand('FS_createPath'); // parent, name, canRead, canWrite
  createCommand('FS_createLazyFile'); // parent, name, canRead, canWrite
  createCommand('FS_createLazyFilesFromList'); // parent, list, parent_url, canRead, canWrite
  createCommand('set_TOTAL_MEMORY'); // size

  var curry = function(obj, fn, args) {
    return function() {
      return obj[fn].apply(obj, args);
    }
  }

  self.compile = function(source_code) {
    var p = new promise.Promise();

    self.compileRaw(source_code).then(function(binary_pdf) {
      if(binary_pdf === false)
        return p.done(false);

      pdf_dataurl = 'data:application/pdf;charset=binary;base64,' + window.btoa(binary_pdf);

      return p.done(pdf_dataurl);
    });
    return p;
  }

  self.compileRaw = function(source_code) {
    if(typeof(chunkSize) === "undefined")
      chunkSize = determineChunkSize();

    var commands;
    if(initialized)
      commands = [
        curry(self, 'FS_unlink', ['/input.tex']),
      ];
    else
      commands = [
        curry(self, 'FS_createFolder', ['/', 'bin', true, true]),
        curry(self, 'FS_createDataFile', ['/bin', 'this.program', '', true, true]),
        curry(self, 'FS_createDataFile', ['/', 'input.tex', source_code, true, true]),
        curry(self, 'FS_createLazyFile', ['/', 'latex.fmt', 'latex.fmt', true, true]),
        curry(self, 'FS_createFolder', ['/bin/', 'share', true, true]),
        curry(self, 'FS_createLazyFile', ['/bin/', 'texmf.cnf', './texlive/texmf-dist/web2c/texmf.cnf', true, true]),
        curry(self, 'FS_createLazyFilesFromList', ['/', 'texlive.lst', './texlive/', true, true]),
      ];

    var sendCompile = function() {
      initialized = true;
      return sendCommand({
        'command': 'run',
        'arguments': ['-interaction=nonstopmode', '-output-format', 'pdf', '&latex', 'input.tex'],
//        'arguments': ['-debug-format', '-output-format', 'pdf', '&latex', 'input.tex'],
      });
    };

    var getPDF = function() {
      console.log(arguments);
      return self.FS_readFile('/input.pdf');
    }

    return promise.chain(commands)
      .then(sendCompile)
      .then(getPDF);
  };
};
